services/wallet/reader.go

```diff
@@ -0,0 +1,238 @@
+package wallet
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/status-im/status-go/multiaccounts/accounts"
+	"github.com/status-im/status-go/services/wallet/chain"
+	"github.com/status-im/status-go/services/wallet/transfer"
+)
+
+func NewReader(s *Service, db *accounts.Database) *Reader {
+	return &Reader{s, db}
+}
+
+type Reader struct {
+	s  *Service
+	db *accounts.Database
+}
+
+type ReaderToken struct {
+	Token         *Token       `json:"token"`
+	OraclePrice   float64      `json:"oraclePrice"`
+	CryptoBalance *hexutil.Big `json:"cryptoBalance"`
+	FiatBalance   *big.Float   `json:"fiatBalance"`
+}
+
+type ReaderAccount struct {
+	Account      *accounts.Account               `json:"account"`
+	Collections  map[uint64][]*OpenseaCollection `json:"collections"`
+	Tokens       map[uint64][]*ReaderToken       `json:"tokens"`
+	Transactions map[uint64][]*transfer.View     `json:"transactions"`
+}
+
+type Wallet struct {
+	Accounts            []*ReaderAccount                 `json:"Accounts"`
+	Favorites           []*Favourite                     `json:"favorites"`
+	OnRamp              []*CryptoOnRamp                  `json:"onRamp"`
+	SavedAddresses      map[uint64][]*SavedAddress       `json:"saveAddresses"`
+	Tokens              map[uint64][]*Token              `json:"tokens"`
+	CustomTokens        []*Token                         `json:"customTokens"`
+	PendingTransactions map[uint64][]*PendingTransaction `json:"pendingTransactions"`
+
+	Currency string `json:"currency"`
+}
+
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func getAddresses(accounts []accounts.Account) []common.Address {
+	addresses := make([]common.Address, len(accounts))
+	for _, account := range accounts {
+		addresses = append(addresses, common.Address(account.Address))
+	}
+	return addresses
+}
+
+func (r *Reader) buildReaderAccount(
+	ctx context.Context,
+	chainIDs []uint64,
+	account accounts.Account,
+	visibleTokens []*Token,
+	prices map[string]float64,
+	balances map[common.Address]*hexutil.Big,
+) (*ReaderAccount, error) {
+	limit := (*hexutil.Big)(big.NewInt(20))
+	toBlock := (*hexutil.Big)(big.NewInt(0))
+
+	collections := make(map[uint64][]*OpenseaCollection)
+	tokens := make(map[uint64][]*ReaderToken)
+	transactions := make(map[uint64][]*transfer.View)
+	for _, chainID := range chainIDs {
+		client, err := newOpenseaClient(chainID, r.s.openseaAPIKey)
+		if err == nil {
+			c, _ := client.fetchAllCollectionsByOwner(common.Address(account.Address))
+			collections[chainID] = c
+		}
+
+		for _, token := range visibleTokens {
+			oraclePrice := prices[token.Symbol]
+			cryptoBalance := balances[token.Address].ToInt()
```

Comment:
Could it be possible that the balance for a token does not exist in the map, and we run into a null exception when calling `.ToInt()`?

Reply:
That shouldn't happen, at least I don t see how, as we are fetching the balances for all addresses/all visible tokens. So should be at least 0